### PR TITLE
add module aws_step_functions_state_machine_execution

### PR DESCRIPTION
**Pull request by prasadkatti**
##### SUMMARY
New module to start or stop execution of an AWS Step functions state machine

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
aws_step_functions_state_machine_execution

##### ADDITIONAL INFORMATION
- The module allows starting or stopping an execution of a state machine. The scope does not include listing or describing executions.
- I will add a PR to the aws-terminator project as well. (https://github.com/mattclay/aws-terminator/pull/71)

### CODE REVIEW

**Comment by prasadkatti** [comment]
ready_for_review

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py]: niggle:

```suggestion
        description: Desired action (start or stop) for a state machine execution.
```

Descriptions should end with a "."
https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-block

File: test/integration/targets/aws_step_functions_state_machine/tasks/main.yml
Comment (tremble) [inline test/integration/targets/aws_step_functions_state_machine/tasks/main.yml]: Niggle 2:

The common AWS connection info can be better written by adding module_defaults to the top level block:

```
- name: Integration test for AWS Step Function state machine module
  module_defaults:
    group/aws:
      aws_access_key: "{{ aws_access_key }}"
      aws_secret_key: "{{ aws_secret_key }}"
      security_token: "{{ security_token | default(omit) }}"
      region: "{{ aws_region }}"
  block:
```

This means you can then remove all of the ``<<: *aws_connection_info`` lines and the "Set connection information for all tasks" task.

**Review by tremble — CHANGES_REQUESTED** [review]
Couple of minor niggles...

File: test/integration/targets/aws_step_functions_state_machine/tasks/main.yml
Comment (tremble) [reply test/integration/targets/aws_step_functions_state_machine/tasks/main.yml]: Oh, and related niggle 3, please add the new module to lib/ansible/config/module_defaults.yml

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py]: You can achieve this same result output with
`module.exit_json(changed=True,**camel_dict_to_snake_dict(execution))`
which can be imported from `ansible.module_utils.ec2`. That will automatically cover if AWS ever starts returning additional keys from this API and is generally easier to maintain.

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py]:
```suggestion
    module.exit_json(changed=True,**camel_dict_to_snake_dict(res)))
```

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: `client.start_execution` is idempotent but `start_execution` always returns `changed=True` if `action=start`. We probably want to `client.list_executions` to see if the execution is already running before proceeding.

File: test/integration/targets/aws_step_functions_state_machine/tasks/main.yml
Comment (jillr) [inline test/integration/targets/aws_step_functions_state_machine/tasks/main.yml:148]: Once the module is idempotent, we'll want another test like `name: Start execution of state machine - idempotent`, then assert that output is not changed.

File: test/integration/targets/aws_step_functions_state_machine/tasks/main.yml
Comment (jillr) [inline test/integration/targets/aws_step_functions_state_machine/tasks/main.yml:199]: And an idempotent stop test.

**Review by jillr — COMMENTED** [review]
Thanks @prasadkatti, this module looks pretty good just a couple things.

File: test/integration/targets/aws_step_functions_state_machine/tasks/main.yml
Comment (prasadkatti) [reply test/integration/targets/aws_step_functions_state_machine/tasks/main.yml]: done. Thanks for the review.

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py]: Thanks for the review. I have pushed a change for this.

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: > The `list_executions` is eventually consistent. The results are best effort and may not reflect very recent updates and changes.

So we could potentially get false positives (changed=True) if the execution was already running but did not get listed by the `list_executions` call.

Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: > StartExecution is idempotent. If StartExecution is called with the same name and input as a running execution, the call will succeed and return the same response as the original request. If the execution is closed or if the input is different, it will return a 400 ExecutionAlreadyExists error. Names can be reused after 90 days.

I return `changed=False` if ExecutionAlreadyExists is returned. That would mean that a new execution was not started. 

But I am not sure if we want to return `changed=False` in the case of a running execution. I guess that really depends on how the value of `changed` is going to be used further down the playbook. Also, if the AWS service isn't making the distinction clear maybe we don't need to either. What do you think?

Comment (jillr) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: I haven't actually used stepfunctions myself so I'm just going off the docs. AIUI, `ExecutionAlreadyExists` would only be returned on a running execution if the `execution_input` is different, not in a case where the input is the same.  As an ansible user who is new to step functions, I feel like if I gave the same input and ran the playbook twice I would expect the 2nd time to be `changed=false` if the first execution is still running.  But that doesn't mean that's how AWS users familiar with step functions will expect it.

Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: We could just document that the value of `changed` for the tasks will be aligned to how boto3 call returns. If we get `ExecutionAlreadyExists`, we return `False`. Otherwise we return `True`. Again, what the right value for `changed` is depends on how the user of the module wishes to use it. We should just document it precisely and then gather feedback from the users as they start using the module. Almost wish there was a thing like beta modules to gather early feedback.

Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: Between ansible and AWS, we should be more aligned with AWS on this one, IMO. If they don't make a distinction between `ExecutionAlreadyExists` and `ExecutionCurrentlyRunning`, then we shouldn't either. This and the fact that `list_executions` is only eventually consistent makes me think the current behavior is what we want.

Comment (jillr) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: That makes sense, and I definitely agree we should very clearly document the behaviour.
@tremble, any thoughts on this approach to idempotency here before we proceed?

Comment (tremble) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: I have a *slight* preference for @jillr's approach, and trying to ensure that check_mode behaves as you might expect.  But it's not a strong enough preference to hold anything up.  The behaviour should certainly be documented though.

Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: So...? What should we do here? Document the behavior clearly? If that is what we plan to do I can push another patch.
Given that with the eventually consistent `list_executions` you could get inconsistent results, do we still want to go with Ansible's definition for idempotency here?

Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: I like @jillr's approach. We should err on the side of trying to keep our AWS modules similar. There are a lot of AWS services that use eventual consistency so we have to work around it as best we can. Using waiters after making API calls can help a little bit, though I'm not sure that it's applicable here (I see that there are no boto3 built-in waiters yet for this service, but you could possibly add your own criteria for waiting to https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/aws/waiters.py). In some circumstances we've added brief pause tasks to integration tests to help. Do add a note in the documentation about the eventual consistency problem.

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:135]: Please fail safely for the module if a BotoCoreError exception happens here too and in stop_execution.

File: test/integration/targets/aws_step_functions_state_machine/tasks/main.yml
Comment (prasadkatti) [reply test/integration/targets/aws_step_functions_state_machine/tasks/main.yml:199]: done 👍

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:135]: fixed, thanks for the review

File: test/integration/targets/aws_step_functions_state_machine/tasks/main.yml
Comment (prasadkatti) [reply test/integration/targets/aws_step_functions_state_machine/tasks/main.yml:148]: added more tests for idempotency and check mode

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:191]: I have pushed more changes. For start execution, we now list all executions and check if there is an execution with the same name. If there is one, the module exits with `changed=False`. For stop execution, we describe the execution and exit with `changed=False` unless the execution status is '_RUNNING_'. I have also added tests around these changes and comments wherever I felt necessary.

**Comment by prasadkatti** [comment]
ready_for_review
Tests failing on AccessDenied to some actions added in https://github.com/mattclay/aws-terminator/pull/80

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py]: Paginators have a `build_full_result` method you can use, which is a little nicer than nested loops. So you could do:
```
current_executions = resp_iterators.build_full_result()['executions']
for execution in current_executions:
    if name == execution['name']:
```

Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py]: Thanks, I will make the required changes. Does it look good otherwise?

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py]: Sorry I missed this before @prasadkatti, did you mean to put a msg here other than 'something'?

**Review by jillr — COMMENTED** [review]
One last small thing that I missed before, then this looks good.

File: lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py
Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py]: Oops. Sorry about that. I was just using that extra return parameter as a way of distinguishing between the 2 `exit_json`s and making sure the right one was called. I will remove it right now.

BTW, the `ExecutionAlreadyExists` block might be unreachable now. But I left in because if that exception does somehow get returned, we don't want the task to fail.

**Review by jillr — APPROVED** [review]
LGTM once terminator change is merged

**Comment by jillr** [comment]
rebuild_merge

**Comment by prasadkatti** [comment]
Hi @jillr, ansibot did not pick up the `rebuild_merge`. Could it be because the PR has `needs_ci_update` and `needs_revision` labels?

This is the log from (https://ansibullbot.eng.ansible.com/cgi-bin/ansibot_status.cgi)

> 2019-12-06 00:34:07,690 INFO skipping: no changes since last run
2019-12-06 00:34:07,691 INFO finished triage for https://github.com/ansible/ansible/pull/64431
2019-12-06 00:34:07,691 INFO starting triage for https://github.com/ansible/ansible/pull/64428
2019-12-06 00:34:08,095 INFO skipping: no changes since last run

**Comment by tremble** [comment]
The needs_revision is because the last test run failed.  Simplest option is probably to rebase and force push to see if the build works now.

**Comment by prasadkatti** [comment]
Closed and reopened the PR to kick off the CI run.

**Comment by prasadkatti** [comment]
The failed tests are not related to the PR - https://app.shippable.com/github/ansible/ansible/runs/152945/112/tests

**Comment by jillr** [comment]
F31 tests have been marked unstable, it's unrelated to this PR. 65671

**Pull request merged**
